src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs

```diff
@@ -188,12 +192,15 @@ protected NodeContext GetNode(string msbuildLocation, string commandLineArgs, in
                     _processesToIgnore.Add(nodeLookupKey);
 
                     // Attempt to connect to each process in turn.
-                    Stream nodeStream = TryConnectToProcess(nodeProcess.Id, 0 /* poll, don't wait for connections */, hostHandshake, clientHandshake);
-                    if (nodeStream != null)
+                    lock (createNodeLock)
```

Comment:
This will serialize all parallel attempts to connect to existing nodes. I'd change it to have one thread per nodeid trying to connect. First one to connect wins and the others back out and relinquish the node. But it might get too hairy.